modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/seedlist/HistoricalSnapshotOrdinals.scala

```diff
@@ -0,0 +1,121 @@
+
+case class HistoricalSnapshotOrdinals(values: NonEmptyList[Range])
+
+case class InvalidOrdinal(ordinal: String) extends NoStackTrace {
+  override def getMessage = s"Invalid ordinal: $ordinal"
+}
+
+case class InvalidRangeFormat(range: List[String]) extends NoStackTrace {
+  override def getMessage = s"Invalid range format: ${range.mkString("-")}"
+}
+
+case class InvalidInput(str: String) extends NoStackTrace {
+  override def getMessage = s"Invalid input: $str"
+}
+
+case class ReversedRange(start: NonNegLong, end: NonNegLong) extends NoStackTrace {
+  override def getMessage =
+    s"Invalid range: '$start-$end'. The start value must not be greater than the end value."
+}
+
+object HistoricalSnapshotOrdinals {
+
+  implicit val optionalHistoricalSnapshotOrdinalsCellDecoder: CellDecoder[Option[HistoricalSnapshotOrdinals]] =
+    CellDecoder.stringDecoder.emap { s =>
+      parseInput(s)
+        .map(reduce)
+        .andThen(_.toNel match {
+          case Some(value) => value.validNel
+          case None        => InvalidInput(s).invalidNel
+        })
+        .map(HistoricalSnapshotOrdinals(_).some)
+        .leftMap { errs =>
+          val msg = errs.map(_.getMessage).toList.mkString("; ")
+
+          new DecoderError(msg = msg)
+        }
+        .toEither
+    }
+
+  private val valueDelimiter = "-"
+  private val entryDelimiter = ";"
+
+  type ValidationResult[A] = ValidatedNel[Throwable, A]
+
+  private def parseInput(input: String): ValidationResult[Seq[Range]] =
+    input
+      .split(entryDelimiter)
+      .toList
+      .map(_.split(valueDelimiter).map(_.trim).toList)
+      .traverse(parseRange)
+
+  private def parseRange(range: List[String]): ValidationResult[Range] =
+    range match {
+      case start :: Nil =>
+        parseOrdinal(start).map(Range(_, NonNegLong.MaxValue))
+      case start :: end :: Nil =>
+        (parseOrdinal(start), parseOrdinal(end)).mapN {
+          case (start, end) =>
+            Range(start = start, end = end)
+        }.andThen {
+          case r if r.valid => r.validNel
+          case r            => ReversedRange(r.start, r.end).invalidNel
+        }
+      case _ =>
+        InvalidRangeFormat(range).invalidNel[Range]
+    }
+
+  private def parseOrdinal(ordinal: String): ValidationResult[NonNegLong] =
+    Either
+      .catchOnly[NumberFormatException](ordinal.toLong)
+      .leftMap(_ => InvalidOrdinal(ordinal))
+      .flatMap { l =>
+        NonNegLong
+          .from(l)
+          .leftMap(_ => InvalidOrdinal(ordinal))
+      }
+      .toValidatedNel
+
+  private def reduce(ranges: Seq[Range]): List[Range] = {
+    val sortedRanges = ranges.sortBy(_.start.value)
+    val mergedRanges = sortedRanges.foldLeft(List.empty[Range]) { (acc, curr) =>
+      acc match {
+        case prev :: tail if prev.overlaps(curr) =>
+          prev.merge(curr) :: tail
+        case _ =>
+          Range(curr.start, curr.end) :: acc
+      }
+    }
```

Comment:
1. you can eliminate the `match` by adjusting the function to `foldLeft`
2. You don't need to create another range: `Range(curr.start, curr.end) == curr`

Reply:
Changed to

```scala
    val mergedRanges = sortedRanges.foldLeft(List.empty[Range]) {
      case (prev :: tail, curr) if prev.overlaps(curr) =>
        prev.merge(curr) :: tail
      case (acc, curr) =>
        curr :: acc
    }
```